File: go/src/otest/runner.go
package otest

import (
	"fmt"
	"time"
)

type Measurement struct {
	N    int
	Time time.Duration
}

func RunN[T any](
	targetFn func(data T),
	dataBakeFn func(n int) T,
	maxN int, points int,
) []Measurement {
	ns := linspace(1, maxN, points)

	res := make([]Measurement, 0, len(ns))

	for _, n := range ns {
		data := dataBakeFn(n)

		start := time.Now()
		targetFn(data)
		elapsed := time.Since(start)

		res = append(res, Measurement{N: n, Time: elapsed})
	}

	return res
}

func RunOTest(name string, fn func() []Measurement, preheat int, repeates int) {
	fmt.Print("\033[s")
	for i := 0; i < preheat; i++ {
		fmt.Printf("\033[u\033[Kpreheating %v: %v/%v", name, i+1, preheat)
		fn()
	}

	RemoveReports(name)
	for i := 0; i < repeates; i++ {
		fmt.Printf("\033[u\033[K%v: %v/%v", name, i+1, repeates)
		m := fn()
		WriteReport(fmt.Sprintf("%v_%v", name, i), m)
	}
	fmt.Println()
}

File: rust/src/otest/runner.rs
use crate::otest::report;

use super::utils::linspace;
use std::{
    io::{stdout, Write},
    time::Duration,
};

pub struct Measurement {
    pub n: usize,
    pub time: Duration,
}

pub fn run_n<T>(
    test_fn: impl Fn(&mut T),
    data_bake_fn: impl Fn(usize) -> T,
    max_n: usize,
    points: usize,
) -> Vec<Measurement> {
    let ns: Vec<usize> = linspace(1, max_n as i32, points as i32)
        .iter()
        .map(|&x| x as usize)
        .collect();

    let mut res = Vec::<Measurement>::with_capacity(ns.len());

    for n in ns {
        let mut data = data_bake_fn(n);

        let start = std::time::Instant::now();
        std::hint::black_box(test_fn(std::hint::black_box(&mut data)));
        let elapsed = start.elapsed();

        res.push(Measurement { n, time: elapsed });
    }

    return res;
}

pub fn run_o_test(
    name: &str,
    test_fn: impl Fn() -> Vec<Measurement>,
    preheat: usize,
    repeates: usize,
) {
    print!("\x1b[s");
    for i in 0..preheat {
        print!("\x1b[u\x1b[Kpreheating {}: {}/{}", name, i + 1, preheat);
        stdout().flush().unwrap();
        test_fn();
    }

    report::remove_reports(name).unwrap();
    for i in 0..repeates {
        print!("\x1b[u\x1b[K{}: {}/{}", name, i + 1, repeates);
        stdout().flush().unwrap();
        let res = test_fn();
        report::write_report(&format!("{}_{}", name, i), res).unwrap();
    }
    println!();
    stdout().flush().unwrap();
}
